### PR TITLE
Reduce amount of allocations and speed up DreamPath

### PR DESCRIPTION
**Pull request by Farseer2**
Should improve #498 and generally somewhat speed up DreamPath

### CODE REVIEW

**Comment by ike709** [comment]
While testing with NTstation this exception is happening _a lot_ (with various DM code), and I'm fairly certain it isn't an issue on the master branch (but correct me if I'm wrong):

```
[ERRO] root: Exception Occured: Unable to cast object of type 'OpenDreamRuntime.Objects.DreamObject' to type 'OpenDreamRuntime.Objects.DreamList'.
=DM StackTrace=
   //load_admin_ranks(...)
   //load_admins(...)
   /world/New(...)

=C# StackTrace=
System.InvalidCastException: Unable to cast object of type 'OpenDreamRuntime.Objects.DreamObject' to type 'OpenDreamRuntime.Objects.DreamList'.
   at OpenDreamRuntime.DreamValue.TryGetValueAsDreamList(DreamList& list) in C:\Users\main\Documents\GitHub\OpenDream\OpenDream\OpenDreamRuntime\DreamValue.cs:line 201
   at OpenDreamRuntime.Objects.MetaObjects.DreamMetaObjectList.OperatorAppend(DreamValue a, DreamValue b) in C:\Users\main\Documents\GitHub\OpenDream\OpenDream\OpenDreamRuntime\Objects\MetaObjects\DreamMetaObjectList.cs:line 63
   at OpenDreamRuntime.Procs.DMOpcodeHandlers.Append(DMProcState state) in C:\Users\main\Documents\GitHub\OpenDream\OpenDream\OpenDreamRuntime\Procs\DMOpcodeHandlers.cs:line 532
   at OpenDreamRuntime.Procs.DMProcState.InternalResume() in C:\Users\main\Documents\GitHub\OpenDream\OpenDream\OpenDreamRuntime\Procs\DMProc.cs:line 0
   at OpenDreamRuntime.ProcState.Resume() in C:\Users\main\Documents\GitHub\OpenDream\OpenDream\OpenDreamRuntime\DreamThread.cs:line 86
   ```

**Comment by wixoaGit** [comment]
That exception sounds related to [this](https://github.com/wixoaGit/OpenDream/pull/499/files#diff-1db1d82c72a8865663cf35a3f11512a379fe771d6794c74ae896f6d884ad2b1cR128) line change. `IsDescendantOf()` is used to determine if an object is a `/list` and thus can be cast into a DreamList.

**Comment by ike709** [comment]
Can confirm that this isn't an issue on the master branch, so it's probably that change.

**Comment by ike709** [comment]
I don't have hard benchmark data but this seems _a lot_ slower just eyeballing NTstation init time. I've been sitting in pregame setup for so long that I'm worried it might've hanged on something.

**Comment by Farseer2** [comment]
Has a big indeed - my bad, didn't trigger in the test game, I'll test again
and fix it tonight

File: OpenDreamShared/Dream/DreamPath.cs [hunk 1]
```diff
@@ -214,25 +213,32 @@ public enum PathType {
...
 
-        private void Normalize()
+        private void Normalize(bool canHaveEmptyEntries)
         {
-            var writeIdx = 0;
-            for (var i = 0; i < _elements.Length; i++)
+            if (canHaveEmptyEntries && _elements.Contains(""))
+            {
+                // Slow path :(
+                _elements = _elements.Where(el => !string.IsNullOrEmpty(el)).ToArray();
```
Comment (ike709) [inline OpenDreamShared/Dream/DreamPath.cs:221]: Please avoid using LINQ

Comment (wixoaGit) [reply OpenDreamShared/Dream/DreamPath.cs:221]: This slow path executes very rarely, I don't think it's a big performance concern.

**Review by wixoaGit — APPROVED** [review]

**Pull request merged**